num/dualquat/dual.go

```diff
@@ -0,0 +1,157 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package dualquat
+
+import (
+	"bytes"
+	"fmt"
+
+	"gonum.org/v1/gonum/num/dual"
+	"gonum.org/v1/gonum/num/quat"
+)
+
+// Number is a float64 precision dual quaternion.
+type Number struct {
+	Real, Dual quat.Number
+}
+
+var (
+	zero     Number
+	zeroQuat quat.Number
+)
+
+// Format implements fmt.Formatter.
+func (d Number) Format(fs fmt.State, c rune) {
+	prec, pOk := fs.Precision()
+	if !pOk {
+		prec = -1
+	}
+	width, wOk := fs.Width()
+	if !wOk {
+		width = -1
+	}
+	switch c {
+	case 'v':
+		if fs.Flag('#') {
+			fmt.Fprintf(fs, "%T{%#v, %#v}", d, d.Real, d.Dual)
+			return
+		}
+		c = 'g'
+		prec = -1
+		fallthrough
+	case 'e', 'E', 'f', 'F', 'g', 'G':
+		fre := fmtString(fs, c, prec, width, false)
+		fim := fmtString(fs, c, prec, width, true)
+		fmt.Fprintf(fs, fmt.Sprintf("(%s+%[2]sϵ)", fre, fim), d.Real, d.Dual)
+	default:
+		fmt.Fprintf(fs, "%%!%c(%T=%[2]v)", c, d)
+		return
+	}
+}
+
+// This is horrible, but it's what we have.
```

Comment:
Almost not though!

Reply:
?

Reply:
1.12 is almost out, which means 1.9 support will be dropped from Gonum

Reply:
Ah. That's not the horrible thing. The horrible thing is that the `fmt.Formatter` API does not allow a pass through without constructing a new format string. I can just say, format these things with the same format as I already have, so we have to perform at least one more allocation and then reparse the format string.